### ISSUE TITLE
Flowchart of pytest test session states and hooks

### ISSUE THREAD

**Issue by Sup3rGeo** [pytest-dev/pytest#3261]
Hi,

This is a documentation request to have a flowchart of all the pytest test session states (setup, conftest, collecting, run setup, run test, run teardown, make report, etc..) together with all the applicable hooks.

Something similar to the one on logging module documentation (I suppose that for pytest this would be way more complex):
https://docs.python.org/2/howto/logging.html#logging-flow

I surely read the page on the hooks but not knowing all the steps that are taken by pytest it becomes a bit hard to guess which one we should be using. This would really improve the understanding on how pytest works, especially for plugin developers.

**Comment by nicoddemus** [comment]
Hi @Sup3rGeo, this would definitely be useful.

I think we can use [sphinx.ext.graphviz](http://build-me-the-docs-please.readthedocs.io/en/latest/Using_Sphinx/UsingGraphicsAndDiagramsInSphinx.html) to generate the flowcharts.

**Comment by nicoddemus** [comment]
Just played around with `sphinx.ext.graphviz` and this code:

```
.. digraph:: foo

    "pytest_cmdline_main" -> "pytest_cmdline_preparse" -> "pytest_load_initial_conftests";

    "pytest_cmdline_main" -> "pytest_cmdline_parse";
```

Produces this image:

![graphviz-45befaaca8f80d2a294cbf4685e5fc9ca1222c57](https://user-images.githubusercontent.com/1085180/36702594-85061c7c-1b36-11e8-9baf-dcae376f6318.png)

So this is definitely doable. 👍

**Comment by Sup3rGeo** [comment]
Nice!

I would be willing to help with documentation (with graphviz) but I would just need someone who knows pytest execution flow to define in any way (text or in a tool like [draw.io](http://www.draw.io) for instance)

**Comment by nicoddemus** [comment]
@Sup3rGeo thanks for the offer, we appreciate it!

The hook order can be obtained by passing `--debug`, which will make pytest write a `pytestdebug.log` file with contents like this:

```
versions pytest-3.4.1.dev39+ga1c3c3d.d20180226, py-1.5.2, python-3.6.3.final.0
cwd=C:\pytest
args=['.tmp\\test_fail.py', '--debug', '-v']

  pytest_cmdline_main [hook]
      config: <_pytest.config.Config object at 0x0000016BDEBF8BE0>
    pytest_plugin_registered [hook]
        plugin: <Session '.tmp'>
        manager: <_pytest.config.PytestPluginManager object at 0x0000016BDDD63F60>
    finish pytest_plugin_registered --> [] [hook]
    pytest_configure [hook]
        config: <_pytest.config.Config object at 0x0000016BDEBF8BE0>
      pytest_plugin_registered [hook]
          plugin: <_pytest.cacheprovider.LFPlugin object at 0x0000016BDEFF4358>
          manager: <_pytest.config.PytestPluginManager object at 0x0000016BDDD63F60>
      finish pytest_plugin_registered --> [] [hook]
      pytest_plugin_registered [hook]
          plugin: <_pytest.logging.LoggingPlugin object at 0x0000016BDEFF4518>
          manager: <_pytest.config.PytestPluginManager object at 0x0000016BDDD63F60>
      finish pytest_plugin_registered --> [] [hook]
    find_module called for: py._io [assertion]
    find_module called for: py._io.terminalwriter [assertion]
    find_module called for: termios [assertion]
    find_module called for: termios [assertion]
      pytest_plugin_registered [hook]
...
```

Which can be used as a starting point, and in any case if there are any questions we would gladly answer them over a WIP PR. 😁

**Comment by Sup3rGeo** [comment]
That is really great, can't believe all this time I didn't know it!

I will try to come up with something then and create PR so you can review it.

**Comment by Sup3rGeo** [comment]
I was thinking about doing something like this:
https://github.com/Sup3rGeo/pytest/blob/ace24468d336fdd4a8c2eaa3c3e9a32da961f98b/doc/en/img/Pytest%20hooks.png

However I am not sure how to do this with graphviz. Mainly to represent the fact that some hooks are executed inside the context of other hooks.

What do you think?

**Comment by nicoddemus** [comment]
> I was thinking about doing something like this
> However I am not sure how to do this with graphviz.

I never used `graphviz` myself 🤕 

> What do you think?

I think it looks great! Also, perhaps you can write a script to generate the `graphviz` output from a `pytestdebug.log` file? Then we can automate this graph generation for each release.

**Comment by Sup3rGeo** [comment]
> I think it looks great! Also, perhaps you can write a script to generate the graphviz output from a pytestdebug.log file? Then we can automate this graph generation for each release.

That is a great idea. From what I could see, not all the hooks are shown in the log file. Is that dependent on what pytest features the test is actually using?
If that's the case, is there any test that actually outputs all the hooks, based on which we can derive the graph?

**Comment by nicoddemus** [comment]
> If that's the case, is there any test that actually outputs all the hooks, based on which we can derive the graph?

Not offhand, but IMHO don't worry about it too much: having the script in place generating a graph with the most common hooks and integrated with our release process is already a huge addition. We can always improve that bit later. 😁

**Comment by Sup3rGeo** [comment]
So basically I wrote something to parse the log file and create a tree using `anytree` module:
https://github.com/Sup3rGeo/pytest/blob/master/doc/pytesthooks.py

The result:
```
root
└── pytest_cmdline_main
    ├── pytest_plugin_registered
    ├── pytest_configure
    │   └── pytest_plugin_registered
    ├── pytest_sessionstart
    │   ├── pytest_plugin_registered
    │   └── pytest_report_header
    ├── pytest_collection
    │   ├── pytest_collectstart
    │   ├── pytest_make_collect_report
    │   │   ├── pytest_collect_file
    │   │   │   └── pytest_pycollect_makemodule
    │   │   └── pytest_pycollect_makeitem
    │   │       └── pytest_generate_tests
    │   │           └── pytest_make_parametrize_id
    │   ├── pytest_collectreport
    │   ├── pytest_itemcollected
    │   ├── pytest_collection_modifyitems
    │   └── pytest_collection_finish
    │       └── pytest_report_collectionfinish
    ├── pytest_runtestloop
    │   └── pytest_runtest_protocol
    │       ├── pytest_runtest_logstart
    │       ├── pytest_runtest_setup
    │       │   └── pytest_fixture_setup
    │       ├── pytest_runtest_makereport
    │       ├── pytest_runtest_logreport
    │       │   └── pytest_report_teststatus
    │       ├── pytest_runtest_call
    │       │   └── pytest_pyfunc_call
    │       ├── pytest_runtest_teardown
    │       │   └── pytest_fixture_post_finalizer
    │       └── pytest_runtest_logfinish
    ├── pytest_sessionfinish
    │   └── pytest_terminal_summary
    └── pytest_unconfigure
```

The challenge was to not repeat hook calls and merge internal hooks from some calls that repeat but not exactly one after the other. But it seems to be working fine.

I faced just one problem though - the last `pytest_runtest_setup [hook]` line did not have a corresponding `finish pytest_runtest_setup`. But I am going to run the test again to see.

It also has the ability to generate dot graphs out of it, although the tree text representation is already very good and this graph is just unreadable:

https://github.com/Sup3rGeo/pytest/blob/master/doc/pytesthooks.png

**Comment by Sup3rGeo** [comment]
So I ran a test again and same thing happened: One of the `pytest_runtest_setup` hooks did not have a corresponding `finish pytest_runtest_setup` line. This will confuse the script I wrote.

One possible solution is to consider the leading whitespaces to do the nesting.

**Comment by Sup3rGeo** [comment]
This might be also useful for #3113 and #2670

**Comment by RonnyPfannschmidt** [comment]
looks fabulous, can we add the scripts somewhere to the pytest repo so we can update in future (maybe it makes sense to put this into pluggy)

**Comment by Sup3rGeo** [comment]
@RonnyPfannschmidt nice to hear that!

Merging repeated calls inside a parent is problematic because it does not capture the fact that, for instance, pytest__runtest_makereport runs three times, after setup, call, and teardown respectively.

On the other hand, if we leave everything as is then typically you can have much useless repetition, for example many many pytest_plugin_registered and then the hooks from pytest_logstart to pytest_logfinish for every test case if you have many.

So we might have to craft a test that will call just the right amount of hooks, so it looks good on the script, or we should perhaps think of other way.

If we do it manually then it is easier to indicate that a certain hook sequence is repeated for every item collected, etc.

**Comment by RonnyPfannschmidt** [comment]
we have a set of tools that use pluggy - all of those could benefit from a automated way to deal with hooks and their nesting ^^

this is a topic where playing and experimenting with collecting and displaying the data might give insights

**Comment by nicoddemus** [comment]
> If we do it manually then it is easier to indicate that a certain hook sequence is repeated for every item collected, etc.

One idea is to manually indicate to the script a set of hooks that should appear only once, because this will vary from project to project. 😁

**Comment by Sup3rGeo** [comment]
Just stumbled upon this presentation which has a similar tree view of the hooks (including annotations where there is repetitions):

http://devork.be/talks/pluggy/pluggy.html

Do you guys know him? Should we contact him and see if he might have a script for that already?

**Comment by nicoddemus** [comment]
Sure we know @flub! 😁 

@flub did you use a script to generate that?

**Comment by nicoddemus** [comment]
Here's the video of the presentation for those interested: https://www.youtube.com/watch?v=zZsNPDfOoHU

**Comment by Sup3rGeo** [comment]
I updated the script to use a list of hooks to merge or to exclude from merging, but other than that I am out of ideas on the best way to do things like, for instance, indicate that `pytest_runtest_makereport` should come after setup,call and teardown. 

We could define it to appear a number of times (in this case 3), or indicate that it should have one after each one of them explicitly. 

But because this is really code dependent, I cannot really think of a nice general way for solving this.

**Comment by nicoddemus** [comment]
> We could define it to appear a number of times (in this case 3), or indicate that it should have one after each one of them explicitly.

Explicitly defining seems like a good solution to me. 👍

**Comment by Sup3rGeo** [comment]
Hi! So this stopped so far because I couldn't figure out:

- best way to treat repeated hook calls. Some of them we would like to keep (makereport after each state - setup, call, teardown) while some are redundant (plugin_registered)
- how to have a comprehensive image of all hooks without manual intervention. Depending on the settings used to run the pytest section from which we process the debug file, some hooks may be called while others not
- It is better to have an option for the debug file to have a simpler view, with only hook calls, then to process in the script.

I tend to think the best approach would be to implement the last point mentioned as a first step (very clear) and then the resulting tree could be processed the way the user wants as a second step (not clear at all, so maybe should be out of scope of this).

**Comment by hectorv** [comment]
I got here because I was _just_ thinking how cool this would be 😄 
By the way, do you guys know why the `pytest_addoption` hook doesn't appear in `pytestdebug.log`? I'm definitely using it in my `conftest.py`.

**Comment by orihabAnyvision** [comment]
This is amazing, even unfinished. Doesnt have to be perfect/include EVERYTHING to be a big help. 10x!

**Comment by MukunthanML** [comment]
> So basically I wrote something to parse the log file and create a tree using `anytree` module:
> https://github.com/Sup3rGeo/pytest/blob/master/doc/pytesthooks.py
> 
> The result:
> 
> ```
> root
> └── pytest_cmdline_main
>     ├── pytest_plugin_registered
>     ├── pytest_configure
>     │   └── pytest_plugin_registered
>     ├── pytest_sessionstart
>     │   ├── pytest_plugin_registered
>     │   └── pytest_report_header
>     ├── pytest_collection
>     │   ├── pytest_collectstart
>     │   ├── pytest_make_collect_report
>     │   │   ├── pytest_collect_file
>     │   │   │   └── pytest_pycollect_makemodule
>     │   │   └── pytest_pycollect_makeitem
>     │   │       └── pytest_generate_tests
>     │   │           └── pytest_make_parametrize_id
>     │   ├── pytest_collectreport
>     │   ├── pytest_itemcollected
>     │   ├── pytest_collection_modifyitems
>     │   └── pytest_collection_finish
>     │       └── pytest_report_collectionfinish
>     ├── pytest_runtestloop
>     │   └── pytest_runtest_protocol
>     │       ├── pytest_runtest_logstart
>     │       ├── pytest_runtest_setup
>     │       │   └── pytest_fixture_setup
>     │       ├── pytest_runtest_makereport
>     │       ├── pytest_runtest_logreport
>     │       │   └── pytest_report_teststatus
>     │       ├── pytest_runtest_call
>     │       │   └── pytest_pyfunc_call
>     │       ├── pytest_runtest_teardown
>     │       │   └── pytest_fixture_post_finalizer
>     │       └── pytest_runtest_logfinish
>     ├── pytest_sessionfinish
>     │   └── pytest_terminal_summary
>     └── pytest_unconfigure
> ```
> 
> The challenge was to not repeat hook calls and merge internal hooks from some calls that repeat but not exactly one after the other. But it seems to be working fine.
> 
> I faced just one problem though - the last `pytest_runtest_setup [hook]` line did not have a corresponding `finish pytest_runtest_setup`. But I am going to run the test again to see.
> 
> It also has the ability to generate dot graphs out of it, although the tree text representation is already very good and this graph is just unreadable:
> 
> https://github.com/Sup3rGeo/pytest/blob/master/doc/pytesthooks.png

Could you please why pytest_plugin_registered is repeated under pytest_sessionstart. I checked pytestdebug.log and repeats same plugins' registeration. Please clarify.

root
└── pytest_cmdline_main
    ├── pytest_plugin_registered
    ├── pytest_configure
    │   └── pytest_plugin_registered  <-------- registers all _pytest plugins, conftest, setup tools plugins
    ├── pytest_sessionstart
    │   ├── pytest_plugin_registered  <-------- again registers all same above _pytest plugins, conftest, setup tools plugins
    │   └── pytest_report_header

**Comment by The-Compiler** [comment]
> Is there a reason why the hooks flow still isn't added to the documentation?

Probably because nobody took the time to clean things up, perhaps annotate them a bit, and add them to the docs (and keep them updated once there are new hooks). PRs welcome!

As another starting point, here is what I have in my company training slides:

![image](https://github.com/pytest-dev/pytest/assets/625793/7a03bed3-430d-4fe6-a084-cfd061399644)

![image](https://github.com/pytest-dev/pytest/assets/625793/c77d52ee-d8dc-4cf9-b0cd-a200cb402a1f)

![image](https://github.com/pytest-dev/pytest/assets/625793/85a12559-a327-4a07-9f32-67bbbc752a3d)

![image](https://github.com/pytest-dev/pytest/assets/625793/c97e8682-d6d7-474e-8f89-59b89f9a18ef)

![image](https://github.com/pytest-dev/pytest/assets/625793/9b3beb15-f12e-4766-9317-99eef6261177)

> Another thing I am confused about. Why is pytest_runtest_makereport called before pytest_runtest_call? I would assume the report is made after the test is executed.

It's called once in the setup phase and once in the call phase (and again in the teardown phase). The script probably falsely hides those as duplicates. Another example why this probably needs some more work with someone actually going through the hook output manually and making it human-readable without leaving crucial details out.

**Comment by Decee1** [comment]
> As another starting point, here is what I have in my company training slides:

That is surely MUCH better than what I had access to before. Trying to learn all these hooks and how it works is pretty tough when you don't have much to go with. 

> Another example why this probably needs some more work with someone actually going through the hook output manually

That's a good point. Hopefully somebody will find time to do it at some point.

**Comment by storenth** [comment]
Looks like there are no actions we need to do, I suggest to close the issue. Pretty [doc](https://docs.pytest.org/en/8.0.x/reference/reference.html) already exist.

**Comment by nhairs-lumin** [comment]
> Looks like there are no actions we need to do, I suggest to close the issue. Pretty [doc](https://docs.pytest.org/en/8.0.x/reference/reference.html) already exist.

Specifically I think @storenth is reffering to the Hooks section located on that page: https://docs.pytest.org/en/8.0.x/reference/reference.html#hooks

**Comment by Stefanhg** [comment]
@storenth I disagree.
The task is not completed and a flow chart is still needed. The amazing pictures that @The-Compiler shared is something I use constantly while developing because the docs are not sufficient. 

I think the actual output that the `pytesthooks.py` file gives is pretty strong, if any improvements should be done it should be either linking the output from `pytesthooks.py` to the actual hooks in docs so it is easier to see the arguments available for that hook but that is only nice-to-have.

I am not so strong in Sphinx yet but it should be possible to:

1. Generate a tmp path
2. Generate a simple passing testcase `def test_pass: pass`
3. Run `pytest --debug`
4. Output `pytesthooks.py` to a file
5. Attach the contents of the file to Sphinx

And then the task could be completed ..Right?

**Comment by The-Compiler** [comment]
As I said above, the raw output generated by that script has various issues which will probably confuse people reading the docs rather than help.

**Comment by jxramos** [comment]
I have done some graph analysis at times for certain projects I've been on. The experience with graphviz is a little poor in my estimation; if you express the raw data instead in a networkx Digraph there's a lot of powerful network analysis you can do once loaded up.

You can also export the graph as a graphml file which is an xml dialect that is very powerful as you can annotate a lot of metadata over the nodes and create rich visualizations and load them into tools such as yEd where you can map metadata to visualization properties
https://networkx.org/documentation/stable/reference/readwrite/generated/networkx.readwrite.graphml.write_graphml.html

they offer a very fluid interactive viewer that's great for introspecting the graph structure in its context viewer allowing you to identify successors and predecessors and what not.

I just jumped back into this hook order debug analysis with a project I'm on where I need to coopt the arguments coming in and beat some logging config done at runtime. I was seeing the deprecation warnings for my use of `pytest_cmdline_preparse` but not finding that hook in the debug output though it is defined in the conftest.py file. I'm not sure why that hook is absent though it definitely has an outcome that's discernable on the file that gets output to the filename I give it.

```
$ grep -En "pytest_cmdline_preparse|pytest_sessionstart"  pytest_out.log | grep -v deprecated
51:    def pytest_cmdline_preparse(config, args):
68:    def pytest_cmdline_preparse(config, args):
561:    pytest_sessionstart [hook]
993:    finish pytest_sessionstart --> [] [hook]
```

**Comment by jxramos** [comment]
what governs whether or not a particular hook contributes to the pytestdebug.log file? Is the expectation that every hook contributes an entry, what would bar a particular hook from being absent?

So far I've seen absence for `pytest_cmdline_preparse` and another user @hectorv mentions `pytest_addoption` being absent.

**Comment by jxramos** [comment]
snooping around, it must be the `config.pluginmanager.enable_tracing()` doing a lot of heavy lifting

https://github.com/pytest-dev/pytest/blob/abb61df22c33c74cced8ef7c41a414576fd6e6e8/src/_pytest/helpconfig.py#L89-L99

https://github.com/pytest-dev/pytest/blob/abb61df22c33c74cced8ef7c41a414576fd6e6e8/src/_pytest/helpconfig.py#L114-L138

yah looks like it points back to the bedrock pluggy infrastructure like folks observed above

(bummer github only expands source urls from the same repo)
https://github.com/pytest-dev/pluggy/blob/f403ace3050c9d1580f705e899a5d3cffc62e2c2/src/pluggy/_manager.py#L484-L507

**Comment by BeniaminK** [comment]
This is what I'm using for developing the custom plugins - Mermaid state diagram:
```
stateDiagram-v2
    state "pytest_addhooks(pluginmanager)" as pytest_addhooks
    state "pytest_addoption(parser, pluginmanager)" as pytest_addoption
    state "pytest_plugin_registered(plugin, plugin_name, manager)" as pytest_plugin_registered
    state "pytest_load_initial_conftests(early_config, parser, args)" as pytest_load_initial_conftests
    state "pytest_collect_directory(path, parent)" as pytest_collect_directory
    state "pytest_ignore_collect(collection_path, path, config)" as pytest_ignore_collect
    state "pytest_collect_file(file_path, path, parent)" as pytest_collect_file
    state "pytest_pycollect_makemodule(module_path, path, parent)" as pytest_pycollect_makemodule
    state "pytest_pycollect_makeitem(collector, name, obj)" as pytest_pycollect_makeitem
    state "pytest_collection_modifyitems(session, config, items)" as pytest_collection_modifyitems
    state "pytest_sessionfinish(session, exitstatus)" as pytest_sessionfinish
    pytest_cmdline_main: pytest_cmdline_main(config)
    pytest_configure: pytest_configure(config)
    pytest_sessionstart: pytest_sessionstart(session)
    pytest_collection: pytest_collection(session)
    pytest_generate_tests: pytest_generate_tests(metafunc)
    pytest_collection_finish: pytest_collection_finish(session)
    pytest_unconfigure: pytest_unconfigure(config)

    [*] --> pytest_addhooks
    pytest_addhooks --> pytest_addoption
    pytest_addoption --> pytest_plugin_registered
    pytest_plugin_registered  --> pytest_load_initial_conftests
    pytest_load_initial_conftests --> pytest_cmdline_main
    pytest_cmdline_main --> pytest_configure
    pytest_configure --> pytest_sessionstart
    pytest_sessionstart --> Collection
    state Collection {
        pytest_collection --> DirectoryCollection : 1..N
        state DirectoryCollection {
            %% pytest_collection --> pytest_collect_directory : 1..N
            pytest_collect_directory --> pytest_ignore_collect : 1..N
            pytest_collect_directory --> pytest_collect_file : 1..N
                
            pytest_collect_file --> pytest_pycollect_makemodule : 1..N
        }
        
        DirectoryCollection --> GenTests : 1..N

        state GenTests {
            pytest_pycollect_makeitem --> pytest_generate_tests
        }

        GenTests --> pytest_collection_modifyitems
        pytest_collection_modifyitems --> pytest_collection_finish
    }
    
    
    Collection --> pytest_sessionfinish
    pytest_sessionfinish --> pytest_unconfigure
    pytest_unconfigure --> [*]

    %% Not included:
    %% Test running (runtest) hooks
    %% Reporting hooks
    %% Debugging/Interaction hooks
    %% pytest_cmdline_parse: pytest_cmdline_parse(pluginmanager, args)
    %% pytest_make_parametrize_id: pytest_make_parametrize_id(config, val, argname)
    %% pytest_markeval_namespace: pytest_markeval_namespace(config)

```

I have not included the following parts:
* Test running (runtest) hooks (https://docs.pytest.org/en/stable/reference/reference.html#test-running-runtest-hooks)
* Reporting hooks (https://docs.pytest.org/en/stable/reference/reference.html#reporting-hooks)
* Debugging/Interaction hooks (https://docs.pytest.org/en/stable/reference/reference.html#debugging-interaction-hooks)

Which I might include if anyone needs it.

`pytest_cmdline_parse` - not called at all

<img width="1225" height="2278" alt="Image" src="https://github.com/user-attachments/assets/8e4ac92f-bcb2-426e-9dc4-9227fd7ef3b2" />

**Comment by zy1o** [comment]
I had a clanker 🤖 build a diagram for flow tool **--> [here](https://zy1o.github.io/pytest-hook-atlas/scenarios/baseline/latest/). <--**

The diagrams are generated from actual pytest runs using pluggy's monitoring. They show what pytest called in what order and what's nested inside.

Considerations/thoughts:
1. Two diagrams: a simplified one grouped by phase (setup & configuration, collection, run-test protocol, session finish), and a detailed one below/on click.
2. Every pytest release from 3.7.2 onward (29 releases so far), selectable from a drop-down. Releases that don't change the flow just get a label that links to the existing diagram. When the flow changes, a new version cluster starts. I will backfill to 6.0-ish.
3. How to read: **hue** = which phase a hook belongs to, **darkness** = how often it was called, log-scaled, **words beneath the name** = `historic`, `firstresult`, repeat counts.
4. All hooks are linked to hookspec documentation - I spot-checked the links by hand on a few diagrams, and they worked.
5. Github CI monitors PyPI on a weekly schedule and regenerates the diagrams when a new release comes out.
6. The flow depends on plugins and the test suite in question. So far I captured two scenarios - they result with a slightly different flow of hooks.
7. Next up: Planning to add pytest-xdist flow for both worker and controller processes. This is not yet there.

I was planning to take a bite at that for at least 10 years now, before this ticket was even filed, and I've been failing for all these years. Sorry for the delay.

**I hope that it's helpful!** Contributions are welcome if anyone wants a plugin covered.

<img width="961" height="800" alt="Image" src="https://github.com/user-attachments/assets/a5cc95fd-3cb0-4b9b-b3d1-0baad2f23f4e" />

<img width="952" height="984" alt="Image" src="https://github.com/user-attachments/assets/69e3582f-0138-42e8-abfe-43752b343909" />
